test/jdk/java/awt/event/KeyEvent/FunctionKeyTest.java

```diff
@@ -39,47 +51,59 @@
  * @key headful
  */
 
-public class FunctionKeyTest {
-    private static FunctionKeyTester frame;
+public final class FunctionKeyTest {
+    private static Frame frame;
     private static Robot robot;
 
-    static volatile boolean keyPressReceived;
-    static volatile boolean keyReleaseReceived;
+    private static final CyclicBarrier keyPress = new CyclicBarrier(2);
+    private static final CyclicBarrier keyRelease = new CyclicBarrier(2);
 
-    static final StringBuilder failures = new StringBuilder();
+    private static final CountDownLatch frameActivated = new CountDownLatch(1);
 
-    private static void testKey(int keyCode, String keyText) {
-        keyPressReceived = false;
-        keyReleaseReceived = false;
+    private static final List<Error> failures = new ArrayList<>(4);
+    private static final AtomicReference<Exception> edtException = new AtomicReference<>();
 
+    private static void testKey(int keyCode, String keyText) throws Exception {
         robot.keyPress(keyCode);
-
-        if (!keyPressReceived) {
-            failures.append(keyText).append(" key press is not received\n");
+        try {
+            keyPress.await(2, SECONDS);
+        } catch (TimeoutException e) {
+            keyPress.reset();
+            failures.add(new Error(keyText + " key press is not received", e));
         }
 
         robot.keyRelease(keyCode);
-
-        if (!keyReleaseReceived) {
-            failures.append(keyText).append(" key release is not received\n");
+        try {
+            keyRelease.await(2, SECONDS);
+        } catch (TimeoutException e) {
+            keyRelease.reset();
+            failures.add(new Error(keyText + " key release is not received", e));
         }
     }
 
     public static void main(String[] args) throws Exception {
         robot = new Robot();
-        robot.setAutoWaitForIdle(true);
```

Comment:
This makes the updated test run slower because the barriers block EDT for a short while until the two threads — EDT and main thread — arrive to the same point.

`robot.waitForIdle` blocks the main thread from arriving to the barrier and *releasing EDT*.